File: Luohgu/【数据结构1-2】二叉树/01.cpp
#include <bits/stdc++.h>
using namespace std;

int arr[1<<8],tree[1<<9];

void build(int l,int r, int idx)
{
    if (l==r)
    {
        tree[idx]=l;
        return;
    }
    int mid,l_win,r_win;
    mid=(l+r)/2;
    build(l,mid,idx*2);
    build(mid+1,r,idx*2+1);
    l_win=tree[2*idx];
    r_win=tree[2*idx+1];
    if (arr[l_win]>arr[r_win]) tree[idx]=l_win;
    else tree[idx]=r_win;
}

int main(){
    int n;
    cin >> n;
    for (int i=0;i<1<<n;++i) cin >> arr[i];
    build(0,(1<<n)-1,1);
    int l_win,r_win;
    l_win=tree[2];
    r_win=tree[3];
    if (arr[l_win]>arr[r_win]) cout << r_win+1 << "\n";
    else cout << l_win+1 << "\n";
}

File: Luohgu/【数据结构1-2】二叉树/01.py
import sys
data=list(map(int , sys.stdin.read().split()))
n=data[0]
data[0]=0
tree=[0]*(4*(1<<n))
def build(idx,l,r):
    if l==r:
        tree[idx]=l
        return
    mid=(l+r)//2
    build(2*idx,l,mid)
    build(2*idx+1,mid+1,r)
    l_win=tree[2*idx]
    r_win=tree[2*idx+1]
    if data[l_win]>data[r_win]:
        tree[idx]=l_win
    else:
        tree[idx]=r_win
build(1,1,1<<n)
left_champion = tree[2]
right_champion = tree[3]

if data[left_champion] > data[right_champion]:
    print(right_champion)
else:
    print(left_champion)